xds/internal/xdsclient/xdslbregistry/converter.go

```diff
@@ -172,6 +204,65 @@ func convertCustomPolicy(typeURL string, s *structpb.Struct) (json.RawMessage, b
 
+type wrrLBConfig struct {
+	EnableOOBLoadReport     bool                           `json:"enableOobLoadReport,omitempty"`
+	OOBReportingPeriod      internalserviceconfig.Duration `json:"oobReportingPeriod,omitempty"`
+	BlackoutPeriod          internalserviceconfig.Duration `json:"blackoutPeriod,omitempty"`
+	WeightExpirationPeriod  internalserviceconfig.Duration `json:"weightExpirationPeriod,omitempty"`
+	WeightUpdatePeriod      internalserviceconfig.Duration `json:"weightUpdatePeriod,omitempty"`
+	ErrorUtilizationPenalty float64                        `json:"errorUtilizationPenalty,omitempty"`
+}
+
+func convertClientSideWRR(cfg *v3clientsideweightedroundrobinpb.ClientSideWeightedRoundRobin) (json.RawMessage, error) {
+	const (
+		defaultEnableOOBLoadReport     = false
+		defaultOOBReportingPeriod      = 10 * time.Second
+		defaultBlackoutPeriod          = 10 * time.Second
+		defaultWeightExpirationPeriod  = 3 * time.Minute
+		defaultWeightUpdatePeriod      = time.Second
+		defaultErrorUtilizationPenalty = float32(1.0)
+	)
+	enableOOBLoadReport := defaultEnableOOBLoadReport
+	if enableOOBLoadReportCfg := cfg.GetEnableOobLoadReport(); enableOOBLoadReportCfg != nil {
+		enableOOBLoadReport = enableOOBLoadReportCfg.GetValue()
+	}
+	oobReportingPeriod := defaultOOBReportingPeriod
+	if oobReportingPeriodCfg := cfg.GetOobReportingPeriod(); oobReportingPeriodCfg != nil {
+		oobReportingPeriod = oobReportingPeriodCfg.AsDuration()
+	}
+	blackoutPeriod := defaultBlackoutPeriod
+	if blackoutPeriodCfg := cfg.GetBlackoutPeriod(); blackoutPeriodCfg != nil {
+		blackoutPeriod = blackoutPeriodCfg.AsDuration()
+	}
+	weightExpirationPeriod := defaultWeightExpirationPeriod
+	if weightExpirationPeriodCfg := cfg.GetBlackoutPeriod(); weightExpirationPeriodCfg != nil {
+		weightExpirationPeriod = weightExpirationPeriodCfg.AsDuration()
+	}
+	weightUpdatePeriod := defaultWeightUpdatePeriod
+	if weightUpdatePeriodCfg := cfg.GetWeightUpdatePeriod(); weightUpdatePeriodCfg != nil {
+		weightUpdatePeriod = weightUpdatePeriodCfg.AsDuration()
+	}
+	errorUtilizationPenalty := defaultErrorUtilizationPenalty
+	if errorUtilizationPenaltyCfg := cfg.GetErrorUtilizationPenalty(); errorUtilizationPenaltyCfg != nil {
+		errorUtilizationPenalty = errorUtilizationPenaltyCfg.GetValue()
+	}
+
+	wrrLBConfig := &wrrLBConfig{
+		EnableOOBLoadReport:     enableOOBLoadReport,
+		OOBReportingPeriod:      internalserviceconfig.Duration(oobReportingPeriod),
+		BlackoutPeriod:          internalserviceconfig.Duration(blackoutPeriod),
+		WeightExpirationPeriod:  internalserviceconfig.Duration(weightExpirationPeriod),
+		WeightUpdatePeriod:      internalserviceconfig.Duration(weightUpdatePeriod),
+		ErrorUtilizationPenalty: float64(errorUtilizationPenalty),
+	}
+
+	lbCfgJSON, err := json.Marshal(wrrLBConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling JSON for type %T: %v", wrrLBConfig, err)
+	}
+	return makeBalancerConfigJSON("weighted_round_robin_experimental", lbCfgJSON), nil
```

Comment:
This can use `weightedroundrobin.Name` for sure.

Reply:
Switched.